File: VFC_Vite/src/lib/ddqn.js
import { LR } from "./constants";

// ── Math helpers ─────────────────────────────────────────────────
function mulberry32(seed) {
  let s = seed;
  return () => {
    s |= 0; s = s + 0x6D2B79F5 | 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = t + Math.imul(t ^ (t >>> 7), 61 | t) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
function heWeights(fanIn, fanOut, seed = 1) {
  const std = Math.sqrt(2 / fanIn), rng = mulberry32(seed);
  return Array.from({ length: fanOut }, () =>
    Array.from({ length: fanIn }, () => (rng() * 2 - 1) * std)
  );
}
function zeros(n) { return new Array(n).fill(0); }
function linear(x, W, b) {
  return b.map((bi, i) => W[i].reduce((s, w, j) => s + w * x[j], bi));
}
function relu(v) { return v.map(x => Math.max(0, x)); }

// ── Dueling DDQN Network ─────────────────────────────────────────
// Implements Equation (21): Q(s,a;θ) = V(s;θ) + [A(s,a;θ) − mean(A)]
export class DuelingDDQN {
  constructor(stateDim, actionDim, hidden = 128, seed = 42) {
    this.stateDim  = stateDim;
    this.actionDim = actionDim;

    // Shared backbone layers
    this.sw1 = heWeights(stateDim, hidden, seed);     this.sb1 = zeros(hidden);
    this.sw2 = heWeights(hidden, hidden, seed + 1);   this.sb2 = zeros(hidden);

    // Value stream V(s) → scalar
    this.vw1 = heWeights(hidden, 64, seed + 2);       this.vb1 = zeros(64);
    this.vw2 = heWeights(64, 1, seed + 3);            this.vb2 = zeros(1);

    // Advantage stream A(s,a) → actionDim values
    this.aw1 = heWeights(hidden, 64, seed + 4);       this.ab1 = zeros(64);
    this.aw2 = heWeights(64, actionDim, seed + 5);    this.ab2 = zeros(actionDim);
  }

  // Forward pass — returns Q, V, A
  forward(state) {
    // Shared
    let h = relu(linear(state, this.sw1, this.sb1));
    h      = relu(linear(h,    this.sw2, this.sb2));
    // Value stream
    let v  = relu(linear(h, this.vw1, this.vb1));
    const V = linear(v, this.vw2, this.vb2)[0];
    // Advantage stream
    let a  = relu(linear(h, this.aw1, this.ab1));
    const A = linear(a, this.aw2, this.ab2);
    const meanA = A.reduce((s, x) => s + x, 0) / A.length;
    // Eq.(21)
    const Q = A.map(ai => V + (ai - meanA));
    return { Q, V, A, meanA };
  }

  // Lightweight SGD on advantage stream weights
  trainStep(state, actionIdx, tdError) {
    const gradScale = 0.01;
    this.aw2 = this.aw2.map((row, i) =>
      row.map((w, j) => w - LR * (-2 * tdError * (state[j] || 0)) * gradScale)
    );
    this.ab2 = this.ab2.map(b => b - LR * (-2 * tdError) * gradScale);
  }

  // Copy weights θ → θ' (target network update)
  copyWeightsFrom(src) {
    const cp = m => m.map(r => [...r]);
    this.sw1 = cp(src.sw1); this.sb1 = [...src.sb1];
    this.sw2 = cp(src.sw2); this.sb2 = [...src.sb2];
    this.vw1 = cp(src.vw1); this.vb1 = [...src.vb1];
    this.vw2 = cp(src.vw2); this.vb2 = [...src.vb2];
    this.aw1 = cp(src.aw1); this.ab1 = [...src.ab1];
    this.aw2 = cp(src.aw2); this.ab2 = [...src.ab2];
  }
}

// ── Experience Replay Buffer ─────────────────────────────────────
// Circular ring buffer storing (s, a, r, s', done) tuples
export class ReplayBuffer {
  constructor(capacity = 512) {
    this.buf = []; this.cap = capacity; this.ptr = 0;
  }
  push(exp) {
    if (this.buf.length < this.cap) this.buf.push(exp);
    else { this.buf[this.ptr] = exp; this.ptr = (this.ptr + 1) % this.cap; }
  }
  sample(n = 32) {
    const out = [];
    for (let i = 0; i < Math.min(n, this.buf.length); i++)
      out.push(this.buf[Math.floor(Math.random() * this.buf.length)]);
    return out;
  }
  get size() { return this.buf.length; }
}

File: VFC_Backend/ddqn.py
"""
Dueling Double Deep Q-Network — PyTorch implementation
Implements Equation (21) from the paper:
Q(s,a;θ) = V(s;θ) + [A(s,a;θ) − (1/|A|)·Σ A(s,a';θ)]
"""

import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
import os
from typing import Tuple, List

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")


class DuelingNetwork(nn.Module):
    """
    Dueling architecture:
      Input → Shared backbone → splits into:
        • Value stream    V(s)    → scalar
        • Advantage stream A(s,a) → action_dim values
      Combined via Eq.(21): Q = V + (A - mean(A))
    """
    def __init__(self, state_dim: int, action_dim: int, hidden: int = 128):
        super().__init__()
        self.action_dim = action_dim

        # Shared backbone — two ReLU layers
        self.shared = nn.Sequential(
            nn.Linear(state_dim, hidden),
            nn.ReLU(),
            nn.Linear(hidden, hidden),
            nn.ReLU(),
        )

        # Value stream V(s) → scalar
        self.value_stream = nn.Sequential(
            nn.Linear(hidden, 64),
            nn.ReLU(),
            nn.Linear(64, 1),
        )

        # Advantage stream A(s,a) → action_dim
        self.advantage_stream = nn.Sequential(
            nn.Linear(hidden, 64),
            nn.ReLU(),
            nn.Linear(64, action_dim),
        )

        # He initialisation — matches paper's training setup
        self._init_weights()

    def _init_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Linear):
                nn.init.kaiming_uniform_(m.weight, nonlinearity="relu")
                nn.init.zeros_(m.bias)

    def forward(self, x: torch.Tensor):
        h = self.shared(x)
        V = self.value_stream(h)            # (batch, 1)
        A = self.advantage_stream(h)        # (batch, action_dim)
        # Equation (21)
        Q = V + (A - A.mean(dim=-1, keepdim=True))
        return Q, V.squeeze(-1), A


class DuelingDDQN:
    """
    Full Dueling DDQN agent with:
    • Online network  (θ)   — trained every step
    • Target network  (θ')  — synced every TARGET_SYNC steps
    • Experience replay buffer
    • ε-greedy exploration
    • Double Q-Learning target — Equation (22)
    • MSE loss               — Equation (23)
    """

    def __init__(
        self,
        state_dim:   int,
        action_dim:  int,
        hidden:      int   = 128,
        lr:          float = 3e-3,
        gamma:       float = 0.99,
        eps_start:   float = 1.0,
        eps_min:     float = 0.05,
        eps_decay:   float = 0.9965,
        target_sync: int   = 60,
        batch_size:  int   = 32,
        buffer_size: int   = 512,
    ):
        self.state_dim   = state_dim
        self.action_dim  = action_dim
        self.gamma       = gamma
        self.eps         = eps_start
        self.eps_min     = eps_min
        self.eps_decay   = eps_decay
        self.target_sync = target_sync
        self.batch_size  = batch_size
        self.step_count  = 0

        # Online & target networks
        self.online = DuelingNetwork(state_dim, action_dim, hidden).to(DEVICE)
        self.target = DuelingNetwork(state_dim, action_dim, hidden).to(DEVICE)
        self.target.load_state_dict(self.online.state_dict())
        self.target.eval()

        # Adam optimizer
        self.optimizer = optim.Adam(self.online.parameters(), lr=lr)
        self.loss_fn   = nn.MSELoss()

        # Replay buffer
        self.buffer: list = []
        self.buffer_ptr   = 0
        self.buffer_size  = buffer_size

    # ── Replay buffer ──────────────────────────────────────────────
    def push(self, s, a, r, s_next, done):
        exp = (s, a, r, s_next, done)
        if len(self.buffer) < self.buffer_size:
            self.buffer.append(exp)
        else:
            self.buffer[self.buffer_ptr] = exp
            self.buffer_ptr = (self.buffer_ptr + 1) % self.buffer_size

    # ── ε-greedy action selection ──────────────────────────────────
    def select_actions(self, state: np.ndarray) -> Tuple[List[int], List[float], float]:
        """
        Returns actions, Q-values, V-value for all action slots.
        Uses online network for inference.
        """
        s_tensor = torch.FloatTensor(state).unsqueeze(0).to(DEVICE)
        with torch.no_grad():
            Q, V, A = self.online(s_tensor)
        q_vals = Q.squeeze(0).cpu().numpy().tolist()
        v_val  = float(V.squeeze(0).cpu().numpy())

        actions = []
        for q in q_vals:
            if np.random.random() < self.eps:
                actions.append(np.random.randint(0, 2))   # explore
            else:
                actions.append(1 if q > 0 else 0)         # exploit
        return actions, q_vals, v_val

    # ── Training step ───────────────────────────────────────────────
    def train_step(self) -> float:
        """
        One gradient update using a random mini-batch.
        Implements Double Q-Learning — Equation (22):
          y = r + γ · Qtarget(s', argmax_a' Qonline(s', a'))
        Loss — Equation (23):
          L(θ) = (y - Q(s,a;θ))²
        """
        if len(self.buffer) < self.batch_size:
            return 0.0

        # Sample mini-batch
        indices = np.random.choice(len(self.buffer), self.batch_size, replace=False)
        batch   = [self.buffer[i] for i in indices]
        S, A, R, S_next, D = zip(*batch)

        S      = torch.FloatTensor(np.array(S)).to(DEVICE)
        A      = torch.LongTensor(A).to(DEVICE)
        R      = torch.FloatTensor(R).to(DEVICE)
        S_next = torch.FloatTensor(np.array(S_next)).to(DEVICE)
        D      = torch.FloatTensor(D).to(DEVICE)

        # Current Q values
        Q_curr, _, _ = self.online(S)
        q_curr = Q_curr.gather(1, A.unsqueeze(1)).squeeze(1)

        # Double Q — Eq.(22): online selects action, target evaluates
        with torch.no_grad():
            Q_next_online, _, _ = self.online(S_next)
            best_actions        = Q_next_online.argmax(dim=1)
            Q_next_target, _, _ = self.target(S_next)
            q_next = Q_next_target.gather(1, best_actions.unsqueeze(1)).squeeze(1)

        y = R + self.gamma * q_next * (1 - D)   # Eq.(22)

        # Loss — Eq.(23)
        loss = self.loss_fn(q_curr, y)

        self.optimizer.zero_grad()
        loss.backward()
        torch.nn.utils.clip_grad_norm_(self.online.parameters(), max_norm=10.0)
        self.optimizer.step()

        self.step_count += 1

        # Sync target network θ' ← θ
        if self.step_count % self.target_sync == 0:
            self.target.load_state_dict(self.online.state_dict())

        # Decay epsilon
        self.eps = max(self.eps_min, self.eps * self.eps_decay)

        return float(loss.item())

    # ── Checkpoint save/load ───────────────────────────────────────
    def save(self, path: str = "checkpoint.pth"):
        torch.save({
            "online":     self.online.state_dict(),
            "target":     self.target.state_dict(),
            "optimizer":  self.optimizer.state_dict(),
            "eps":        self.eps,
            "step_count": self.step_count,
        }, path)
        print(f"[DDQN] Saved checkpoint → {path}")

    def load(self, path: str = "checkpoint.pth"):
        if not os.path.exists(path):
            print(f"[DDQN] No checkpoint found at {path}, starting fresh.")
            return
        try:
            ckpt = torch.load(path, map_location=DEVICE)
            self.online.load_state_dict(ckpt["online"])
            self.target.load_state_dict(ckpt["target"])
            self.optimizer.load_state_dict(ckpt["optimizer"])
            self.eps        = ckpt.get("eps",        self.eps)
            self.step_count = ckpt.get("step_count", 0)
            print(f"[DDQN] Loaded checkpoint from {path} (step={self.step_count}, ε={self.eps:.3f})")
        except (RuntimeError, KeyError) as e:
            print(f"[DDQN] Checkpoint at {path} is incompatible with current network architecture.")
            print(f"[DDQN] Reason: {e}")
            print(f"[DDQN] Starting fresh with random weights.")
